torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2232,10 +2242,81 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::gather(
 }
 
 c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::scatter(
-    std::vector<at::Tensor>& /* unused */,
-    std::vector<std::vector<at::Tensor>>& /* unused */,
-    const ScatterOptions& /* unused */) {
-  TORCH_CHECK(false, "ProcessGroupNCCL does not support scatter");
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<std::vector<at::Tensor>>& inputTensors,
+    const ScatterOptions& opts) {
+
+  static auto invalidArgument = [](const std::string& msg) {
+    TORCH_CHECK(false, "ProcessGroupNCCL::scatter: " + msg);
+  };
+
+  assertRootRank(invalidArgument, opts.rootRank, size_);
+  check_gpu_tensors(outputTensors);
+  assertSingleElementInput(invalidArgument, outputTensors);
+
+  std::vector<at::Tensor> inputFlattened;
+
+  // @lint-ignore CLANGTIDY
+  auto tensor = outputTensors.back();
+  RECORD_PARAM_COMMS(
+      rank_,                    // rank
+      "scatter",                // colName
+      tensor.numel(),           // inSize
+      tensor.numel() *
+        this->getSize(),        // outSize
+      tensor.scalar_type(),     // dType
+      std::vector<int64_t>(),   // inSplitSizes
+      std::vector<int64_t>());  // outSplitSize
+
+  if (getRank() == opts.rootRank) {
+    if (inputTensors.size() != 1) {
+      std::stringstream ss;
+      ss << "requires a single-element input list containing a list with "
+         << getSize() << " tensors.";
+      invalidArgument(ss.str());
+    } else if (inputTensors[0].size() != static_cast<size_t>(getSize())) {
+      std::stringstream ss;
+      ss << "Incorrect input list size " << inputTensors[0].size()
+         << ". Input list size should be " << getSize()
+         << ", same as size of the process group.";
+      invalidArgument(ss.str());
+    }
+
+    inputFlattened = flatten_for_scatter_gather(inputTensors, outputTensors, size_, true);
+    check_gpu_tensors(inputFlattened);
```

Comment:
It seems like we might have unnecessary copies here. Do we really need to create a flattened tensor for scatter? It seems like we anyways do `inputs[r]` in the scatter implementation, so why can't we just pass in a list of tensors to `scatter` and don't do any of the flattening or copying?

Reply:
Also, doesn't this apply even to our `gather` implementation where we can avoid flattening and extra copies?

Reply:
Yeah true, I think it's a legacy issue that most of our utils function built on the top of [`collective`](https://github.com/pytorch/pytorch/blob/82c5f298edf3aa450c491fa480278d0ccdd3891e/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1332), where it only takes a input tensor list and a output tensor list (the length must be the same), we utilize bunch of its checks and work initialization there, and we also do [flattening/copying in MPI impls](https://github.com/pytorch/pytorch/blob/82c5f298edf3aa450c491fa480278d0ccdd3891e/torch/csrc/distributed/c10d/ProcessGroupMPI.cpp#L632). But I do agree with you that most of these copies and flattening aren't necessary and introduce extra cost, let me see what I can do here, might need to refactor NCCL utils to avoid those copying